crates/sui-indexer-alt-framework/src/handlers/cp_mapping.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Arc;
+
+use crate::pipeline::{concurrent::Handler, Processor};
+use crate::schema::cp_mapping;
+use anyhow::Result;
+use diesel::prelude::*;
+use diesel_async::RunQueryDsl;
+use sui_field_count::FieldCount;
+use sui_pg_db::{self as db, Connection};
+use sui_types::full_checkpoint_content::CheckpointData;
+
+#[derive(Insertable, Selectable, Queryable, Debug, Clone, FieldCount)]
+#[diesel(table_name = cp_mapping)]
+pub(crate) struct StoredCpMapping {
+    pub cp: i64,
+    pub tx_lo: i64,
+    pub tx_hi: i64,
+    pub epoch: i64,
+}
+
+pub struct CheckpointMapping {
+    from: StoredCpMapping,
+    to: StoredCpMapping,
+}
+
+pub(crate) struct CpMapping;
+
+impl CheckpointMapping {
+    /// Gets the tx and epoch mappings for both the start and end checkpoints.
+    ///
+    /// The values are expected to exist since the cp_mapping table must have enough information to
+    /// encompass the retention of other tables.
+    pub async fn get_range(
+        conn: &mut Connection<'_>,
+        from_cp: u64,
+        to_cp: u64,
+    ) -> QueryResult<CheckpointMapping> {
+        let results = cp_mapping::table
+            .select(StoredCpMapping::as_select())
+            .filter(cp_mapping::cp.eq_any([from_cp as i64, to_cp as i64]))
+            .order(cp_mapping::cp.asc())
+            .load::<StoredCpMapping>(conn)
+            .await?;
+
+        match results.as_slice() {
+            [first, .., last] => Ok(CheckpointMapping {
+                from: first.clone(),
+                to: last.clone(),
+            }),
+            _ => Err(diesel::result::Error::NotFound),
+        }
+    }
+
+    /// Inclusive start and exclusive end range of checkpoints.
+    pub fn checkpoint_interval(&self) -> (u64, u64) {
+        (self.from.cp as u64, (self.to.cp - 1) as u64)
+    }
+
+    /// Inclusive start and exclusive end range of txs. Because the tx_lo in db is inclusive but
+    /// tx_hi is exclusive, returns None when encountering an empty range (where tx_lo equals
+    /// tx_hi).
+    pub fn tx_interval(&self) -> Option<(u64, u64)> {
+        if self.from.tx_lo == self.to.tx_hi {
+            None
+        } else {
+            Some((self.from.tx_lo as u64, self.to.tx_hi as u64))
+        }
+    }
+
+    /// Inclusive start and exclusive end range of epochs.
```

Comment:
Same here. Why do we need to subtract 1 if it's exclusive? I think it makes more sense to just return `self.to.epoch` here and `self.to.cp` above, to be consistent with `tx_hi` and our convention in the code base.